// @tinystacks/aws-cdk-constructs/dist/constructs/networking/alb.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ALB = void 0;
const constructs_1 = require("constructs");
const elbv2 = __importStar(require("aws-cdk-lib/aws-elasticloadbalancingv2"));
const iac_utils_1 = require("@tinystacks/iac-utils");
const aws_cdk_lib_1 = require("aws-cdk-lib");
class ALB extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const { vpc, albName, internetAccess, securityGroup, albVpcSubnets, clusterName, stackName } = props;
        // create alb
        this._alb = new elbv2.ApplicationLoadBalancer(this, (0, iac_utils_1.constructId)('lb'), {
            vpc: vpc,
            loadBalancerName: albName,
            internetFacing: internetAccess,
            securityGroup: securityGroup,
            vpcSubnets: albVpcSubnets
        });
        // tag
        aws_cdk_lib_1.Tags.of(this).add('ingress.k8s.aws/resource', 'LoadBalancer');
        aws_cdk_lib_1.Tags.of(this).add('elbv2.k8s.aws/cluster', clusterName);
        aws_cdk_lib_1.Tags.of(this).add('ingress.k8s.aws/stack', stackName);
    }
    get albName() {
        return this._alb.loadBalancerName;
    }
}
exports.ALB = ALB;
