// @tinystacks/aws-cdk-constructs/dist/constructs/networking/tagging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubnetTagging = void 0;
const constructs_1 = require("constructs");
const custom_resources_1 = require("aws-cdk-lib/custom-resources");
class SubnetTagging extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const { ec2ResourceTagsRequest } = props;
        const sdkCall = {
            service: 'EC2',
            action: 'createTags',
            parameters: ec2ResourceTagsRequest,
            physicalResourceId: custom_resources_1.PhysicalResourceId.of(id)
        };
        const resourcePolicy = custom_resources_1.AwsCustomResourcePolicy.fromSdkCalls({
            resources: custom_resources_1.AwsCustomResourcePolicy.ANY_RESOURCE
        });
        new custom_resources_1.AwsCustomResource(this, 'tagResources', {
            onUpdate: sdkCall,
            onCreate: sdkCall,
            policy: resourcePolicy
        });
    }
}
exports.SubnetTagging = SubnetTagging;
