// @tinystacks/aws-cdk-constructs/dist/constructs/networking/vpc-peer-dns-resolution.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcPeerDnsResolution = void 0;
const custom_resources_1 = require("aws-cdk-lib/custom-resources");
class VpcPeerDnsResolution extends custom_resources_1.AwsCustomResource {
    constructor(scope, id, props) {
        const { peeringConnectionId, vpcArn, accountId, region, isRequester, isAccepter } = props;
        const enableDnsResolution = {
            service: 'EC2',
            action: 'modifyVpcPeeringConnectionOptions',
            parameters: {
                VpcPeeringConnectionId: peeringConnectionId
            },
            physicalResourceId: custom_resources_1.PhysicalResourceId.of(id)
        };
        const disableDnsResolution = {
            service: 'EC2',
            action: 'modifyVpcPeeringConnectionOptions',
            parameters: {
                VpcPeeringConnectionId: peeringConnectionId
            }
        };
        if (isRequester) {
            enableDnsResolution.parameters.RequesterPeeringConnectionOptions = {
                AllowDnsResolutionFromRemoteVpc: true
            };
            disableDnsResolution.parameters.RequesterPeeringConnectionOptions = {
                AllowDnsResolutionFromRemoteVpc: false
            };
        }
        if (isAccepter) {
            enableDnsResolution.parameters.AccepterPeeringConnectionOptions = {
                AllowDnsResolutionFromRemoteVpc: true
            };
            disableDnsResolution.parameters.AccepterPeeringConnectionOptions = {
                AllowDnsResolutionFromRemoteVpc: false
            };
        }
        const policy = custom_resources_1.AwsCustomResourcePolicy.fromSdkCalls({
            resources: [
                vpcArn,
                `arn:aws:ec2:${region}:${accountId}:vpc-peering-connection/*`
            ]
        });
        super(scope, id, {
            onCreate: enableDnsResolution,
            onUpdate: enableDnsResolution,
            onDelete: disableDnsResolution,
            policy
        });
    }
}
exports.VpcPeerDnsResolution = VpcPeerDnsResolution;
