// @tinystacks/aws-cdk-constructs/dist/constructs/networking/vpc-peering-request-accepter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcPeeringRequestAccepter = void 0;
const custom_resources_1 = require("aws-cdk-lib/custom-resources");
class VpcPeeringRequestAccepter extends custom_resources_1.AwsCustomResource {
    constructor(scope, id, props) {
        const { vpcArn, peeringConnectionId, accountId, region } = props;
        const acceptPeeringConnection = {
            service: 'EC2',
            action: 'acceptVpcPeeringConnection',
            parameters: {
                VpcPeeringConnectionId: peeringConnectionId
            },
            physicalResourceId: custom_resources_1.PhysicalResourceId.of(id)
        };
        const policy = custom_resources_1.AwsCustomResourcePolicy.fromSdkCalls({
            resources: [
                vpcArn,
                `arn:aws:ec2:${region}:${accountId}:vpc-peering-connection/*`
            ]
        });
        super(scope, id, {
            onCreate: acceptPeeringConnection,
            onUpdate: acceptPeeringConnection,
            policy
        });
    }
}
exports.VpcPeeringRequestAccepter = VpcPeeringRequestAccepter;
