// @tinystacks/aws-cdk-constructs/dist/constructs/proxies/alb.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Alb = void 0;
const ec2 = __importStar(require("aws-cdk-lib/aws-ec2"));
const cdk = __importStar(require("aws-cdk-lib"));
const elbv2 = __importStar(require("aws-cdk-lib/aws-elasticloadbalancingv2"));
const constructs_1 = require("constructs");
const iac_utils_1 = require("@tinystacks/iac-utils");
const securitygroups_1 = require("../networking/securitygroups");
class Alb extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const albSecurityGroupRules = [
            { name: 'Internet to ALB', port: ec2.Port.tcp(80), peer: ec2.Peer.anyIpv4() }
        ];
        const { albSecurityGroup = new securitygroups_1.SecurityGroups(this, (0, iac_utils_1.constructId)('alb', 'SecurityGroup'), {
            vpc: props.vpc,
            securityGroupName: 'albSecurityGroup',
            securityGroupRulesList: albSecurityGroupRules
        }).securityGroup } = props;
        this.alb = new elbv2.ApplicationLoadBalancer(this, (0, iac_utils_1.constructId)('alb'), {
            vpc: props.vpc,
            vpcSubnets: { subnets: props.vpc.publicSubnets },
            internetFacing: true
        });
        this.alb.addSecurityGroup(albSecurityGroup);
        this._albTargetGroup = new elbv2.ApplicationTargetGroup(this, (0, iac_utils_1.constructId)('alb', 'TargetGroup'), {
            port: props.applicationPort,
            vpc: props.vpc,
            protocol: elbv2.ApplicationProtocol.HTTP,
            targetType: elbv2.TargetType.IP
        });
        this._albTargetGroup.configureHealthCheck({
            path: props.healthCheckPath,
            protocol: elbv2.Protocol.HTTP
        });
        const albListener = this.alb.addListener((0, iac_utils_1.constructId)('alb', 'Listener'), {
            open: true,
            port: 80
        });
        albListener.addTargetGroups((0, iac_utils_1.constructId)('alb', 'Listener', 'TargetGroup'), {
            targetGroups: [this._albTargetGroup]
        });
        new cdk.CfnOutput(this, (0, iac_utils_1.constructId)('alb', 'DnsName'), {
            value: this.alb.loadBalancerDnsName
        });
    }
    get albTargetGroup() {
        return this._albTargetGroup;
    }
    getAlb() {
        return this.alb;
    }
}
exports.Alb = Alb;
