// @tinystacks/aws-cdk-constructs/dist/constructs/networking/vpc-peering-routes/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcPeeringRoutes = void 0;
const iac_utils_1 = require("@tinystacks/iac-utils");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const aws_lambda_1 = require("aws-cdk-lib/aws-lambda");
const aws_logs_1 = require("aws-cdk-lib/aws-logs");
const custom_resources_1 = require("aws-cdk-lib/custom-resources");
const constructs_1 = require("constructs");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class VpcPeeringRoutes extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const { accountId, region } = props;
        const role = new aws_iam_1.Role(this, (0, iac_utils_1.constructId)(id, 'role'), {
            assumedBy: new aws_iam_1.ServicePrincipal('lambda.amazonaws.com'),
            inlinePolicies: {
                functionPolicy: new aws_iam_1.PolicyDocument({
                    statements: [
                        new aws_iam_1.PolicyStatement({
                            actions: [
                                'ec2:DescribeRouteTables'
                            ],
                            resources: ['*']
                        }),
                        new aws_iam_1.PolicyStatement({
                            actions: [
                                'ec2:CreateRoute',
                                'ec2:DeleteRoute'
                            ],
                            resources: [`arn:aws:ec2:${region}:${accountId}:route-table/*`]
                        })
                    ]
                })
            },
            managedPolicies: [
                {
                    managedPolicyArn: 'arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole'
                }
            ]
        });
        const fn = new aws_lambda_1.Function(this, (0, iac_utils_1.constructId)(id, 'lambdaFunction'), {
            code: new aws_lambda_1.InlineCode(fs_1.default.readFileSync(path_1.default.resolve(__dirname, './lambda.js'), { encoding: 'utf-8' })),
            handler: 'index.handler',
            timeout: aws_cdk_lib_1.Duration.seconds(300),
            runtime: aws_lambda_1.Runtime.NODEJS_16_X,
            role,
            logRetention: aws_logs_1.RetentionDays.THREE_MONTHS
        });
        const provider = new custom_resources_1.Provider(this, 'Provider', {
            onEventHandler: fn
        });
        const resource = new aws_cdk_lib_1.CustomResource(this, 'Resource', {
            serviceToken: provider.serviceToken,
            properties: props
        });
        this.response = resource.getAtt('Response').toString();
    }
}
exports.VpcPeeringRoutes = VpcPeeringRoutes;
