// @tinystacks/aws-cdk-constructs/dist/constructs/compute/ecs-service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcsService = void 0;
const ecs = __importStar(require("aws-cdk-lib/aws-ecs"));
const iam = __importStar(require("aws-cdk-lib/aws-iam"));
const constructs_1 = require("constructs");
const iac_utils_1 = require("@tinystacks/iac-utils");
const lodash_1 = require("lodash");
class EcsService extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const ecsTaskRole = new iam.Role(this, (0, iac_utils_1.constructId)('ecs', 'TaskRole'), {
            assumedBy: new iam.ServicePrincipal('ecs-tasks.amazonaws.com'),
            description: 'Role that the api task definitions use'
        });
        if (!(0, lodash_1.isEmpty)(props.ecsIamPolicyStatements)) {
            ecsTaskRole.attachInlinePolicy(new iam.Policy(this, (0, iac_utils_1.constructId)('ecs', 'TaskPolicy'), {
                statements: props.ecsIamPolicyStatements
            }));
        }
        const ecsTaskDefinition = new ecs.TaskDefinition(this, (0, iac_utils_1.constructId)('ecs', 'TaskDefinition'), {
            family: 'task',
            compatibility: ecs.Compatibility.EC2_AND_FARGATE,
            cpu: String(props.cpu),
            memoryMiB: String(props.memoryLimitMiB),
            networkMode: ecs.NetworkMode.AWS_VPC,
            taskRole: ecsTaskRole
        });
        const ecsContainer = ecsTaskDefinition.addContainer((0, iac_utils_1.constructId)('ecs', 'Container'), {
            containerName: props.containerName,
            image: props.repositoryImage || ecs.RepositoryImage.fromRegistry(props.containerImage || ''),
            memoryLimitMiB: props.memoryLimitMiB,
            environment: props.ecsTaskEnvVars,
            logging: ecs.LogDriver.awsLogs({ streamPrefix: props.containerName }),
            command: props.command
        });
        ecsContainer.addPortMappings({ containerPort: props.applicationPort });
        this.ecsService = new ecs.FargateService(this, (0, iac_utils_1.constructId)('ecs', 'FargfateService'), {
            cluster: props.ecsCluster,
            desiredCount: props.desiredCount,
            taskDefinition: ecsTaskDefinition,
            securityGroups: [props.ecsSecurityGroup],
            assignPublicIp: true,
            enableExecuteCommand: true
        });
        if (props.albTargetGroup) {
            this.ecsService.attachToApplicationTargetGroup(props.albTargetGroup);
        }
    }
    service() {
        return this.ecsService;
    }
}
exports.EcsService = EcsService;
