// @tinystacks/aws-cdk-constructs/dist/constructs/compute/s3.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3 = void 0;
const constructs_1 = require("constructs");
const s3Deployment = __importStar(require("aws-cdk-lib/aws-s3-deployment"));
const s3 = __importStar(require("aws-cdk-lib/aws-s3"));
class S3 extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const { existingBucketArn, bucketProps } = props;
        this.id = id;
        this.bucketProps = bucketProps;
        if (existingBucketArn !== undefined) {
            this.bucket = s3.Bucket.fromBucketArn(this, `${id}-bucket`, existingBucketArn);
        }
        else {
            this.bucket = this.createBucket();
        }
    }
    createBucket() {
        return new s3.Bucket(this, `${this.id}-bucket`, this.bucketProps);
    }
    uploadSource(source, bucketKeyPrefix) {
        new s3Deployment.BucketDeployment(this, `${this.id}-fileUpload`, {
            sources: source,
            prune: false,
            destinationBucket: this.bucket,
            destinationKeyPrefix: bucketKeyPrefix
        });
    }
    get bucketArn() {
        return this.bucket.bucketArn;
    }
    get bucketName() {
        return this.bucket.bucketName;
    }
}
exports.S3 = S3;
