// @tinystacks/aws-cdk-constructs/dist/constructs/compute/eks-cleanup/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EksCleanup = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const constructs_1 = require("constructs");
const iac_utils_1 = require("@tinystacks/iac-utils");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const aws_lambda_1 = require("aws-cdk-lib/aws-lambda");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const custom_resources_1 = require("aws-cdk-lib/custom-resources");
const aws_logs_1 = require("aws-cdk-lib/aws-logs");
class EksCleanup extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const role = new aws_iam_1.Role(this, (0, iac_utils_1.constructId)('EksCleanupRole'), {
            assumedBy: new aws_iam_1.ServicePrincipal('lambda.amazonaws.com'),
            inlinePolicies: {
                functionPolicy: new aws_iam_1.PolicyDocument({
                    statements: [
                        new aws_iam_1.PolicyStatement({
                            actions: [
                                'ec2:DescribeNetworkInterfaces',
                                'ec2:DeleteNetworkInterface',
                                'ec2:DescribeSecurityGroups',
                                'ec2:DeleteSecurityGroup'
                            ],
                            resources: ['*']
                        })
                    ]
                })
            },
            managedPolicies: [
                {
                    managedPolicyArn: 'arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole'
                }
            ]
        });
        const fn = new aws_lambda_1.Function(this, (0, iac_utils_1.constructId)('EksCleanupLambdaFunction'), {
            code: new aws_lambda_1.InlineCode(fs.readFileSync(path.resolve(__dirname, './lambda.js'), { encoding: 'utf-8' })),
            handler: 'index.handler',
            timeout: aws_cdk_lib_1.Duration.seconds(300),
            runtime: aws_lambda_1.Runtime.NODEJS_16_X,
            role,
            logRetention: aws_logs_1.RetentionDays.THREE_MONTHS
        });
        const provider = new custom_resources_1.Provider(this, 'EksCleanupProvider', {
            onEventHandler: fn
        });
        const resource = new aws_cdk_lib_1.CustomResource(this, 'EksCleanupResource', {
            serviceToken: provider.serviceToken,
            properties: props
        });
        this.response = resource.getAtt('Response').toString();
    }
}
exports.EksCleanup = EksCleanup;
