// @tinystacks/aws-cdk-constructs/dist/constructs/db/rds.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rds = void 0;
const cdk = __importStar(require("aws-cdk-lib"));
const rds = __importStar(require("aws-cdk-lib/aws-rds"));
const constructs_1 = require("constructs");
const iac_utils_1 = require("@tinystacks/iac-utils");
class Rds extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const { databasePort = 5432, databaseEngine = rds.DatabaseInstanceEngine.POSTGRES, databaseName = 'tstesting', databaseUsername = 'postgres' } = props;
        this._databaseName = databaseName;
        this._databaseUsername = databaseUsername;
        if (!props.isImported || !props.dbArn) {
            this.RdsInstance = new rds.DatabaseInstance(this, (0, iac_utils_1.constructId)('rds', 'instance'), {
                engine: databaseEngine,
                vpc: props.vpc,
                vpcSubnets: {
                    subnetType: props.subnetType
                },
                instanceType: props.instanceType,
                allocatedStorage: props.storageSize || 20,
                maxAllocatedStorage: (props.storageSize || 20) * 2,
                allowMajorVersionUpgrade: true,
                instanceIdentifier: props.instanceIdentifier,
                databaseName: this._databaseName,
                port: databasePort,
                securityGroups: props.securityGroupsList
            });
            new cdk.CfnOutput(this, 'db-secret', {
                value: `${props.instanceIdentifier}-db-secret:${this.RdsInstance.secret?.secretArn}`
            });
            new cdk.CfnOutput(this, 'db-endpoint-port', {
                value: `${this.RdsInstance.dbInstanceEndpointAddress}:${this.RdsInstance.dbInstanceEndpointPort}`
            });
            new cdk.CfnOutput(this, 'db-name', {
                value: `${this._databaseName}`
            });
            const dbSecretArnOutputId = Rds.OutputDescriptions.secretArn(props.instanceIdentifier);
            new cdk.CfnOutput(this, dbSecretArnOutputId, {
                description: dbSecretArnOutputId,
                value: this.RdsInstance.secret?.secretArn || ''
            });
        }
        else {
            const identifier = props.dbArn.split('db:')[1];
            this.RdsInstance = rds.DatabaseInstance.fromDatabaseInstanceAttributes(this, (0, iac_utils_1.constructId)('postgres'), {
                instanceIdentifier: identifier,
                instanceEndpointAddress: '',
                securityGroups: props.securityGroupsList,
                port: databasePort
            });
        }
    }
    get dbSecret() {
        return this.RdsInstance.secret;
    }
    get db() {
        return this.RdsInstance;
    }
    get dbName() {
        return this._databaseName;
    }
    get dbUsername() {
        return this._databaseUsername;
    }
}
exports.Rds = Rds;
Rds.OutputDescriptions = {
    secretArn(instanceIdentifier) {
        return `${instanceIdentifier}-secret-arn`;
    }
};
