// @tinystacks/iac-utils/dist/naming/naming-restrictions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resourceNameRules = exports.alphaNumericOnlyAllowedCharacters = exports.alphaNumericOnly = exports.alphaNumericAndHyphenAllowedCharacterTypes = exports.alphaNumericAndHyphen = void 0;
const alphaNumericAndHyphen = /[^a-zA-Z0-9-]+/;
exports.alphaNumericAndHyphen = alphaNumericAndHyphen;
const alphaNumericAndHyphenAllowedCharacterTypes = 'alphanumeric characters and hyphens';
exports.alphaNumericAndHyphenAllowedCharacterTypes = alphaNumericAndHyphenAllowedCharacterTypes;
const alphaNumericOnly = /[^a-zA-Z0-9]+/;
exports.alphaNumericOnly = alphaNumericOnly;
const alphaNumericOnlyAllowedCharacters = 'alphanumeric characters';
exports.alphaNumericOnlyAllowedCharacters = alphaNumericOnlyAllowedCharacters;
const resourceNameRules = {
    AWS: {
        s3BucketName: {
            // source: https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html
            characterLimit: 63,
            characterRestrictions: alphaNumericAndHyphen,
            allowedCharacterTypes: alphaNumericAndHyphenAllowedCharacterTypes
        },
        iamRoleName: {
            characterLimit: 64,
            characterRestrictions: alphaNumericAndHyphen,
            allowedCharacterTypes: alphaNumericAndHyphenAllowedCharacterTypes
        },
        lambdaFunctionName: {
            characterLimit: 64,
            characterRestrictions: alphaNumericAndHyphen,
            allowedCharacterTypes: alphaNumericAndHyphenAllowedCharacterTypes
        },
        secretName: {
            characterLimit: 512,
            characterRestrictions: alphaNumericAndHyphen,
            allowedCharacterTypes: alphaNumericAndHyphenAllowedCharacterTypes
        },
        cloudFormationLogicalId: {
            characterLimit: 255,
            characterRestrictions: alphaNumericOnly,
            allowedCharacterTypes: alphaNumericOnlyAllowedCharacters
        },
        apigName: {
            characterLimit: 255,
            characterRestrictions: alphaNumericAndHyphen,
            allowedCharacterTypes: alphaNumericAndHyphenAllowedCharacterTypes
        },
        codebuildProjectName: {
            characterLimit: 255,
            characterRestrictions: alphaNumericAndHyphen,
            allowedCharacterTypes: alphaNumericAndHyphenAllowedCharacterTypes
        },
        logGroupName: {
            characterLimit: 512,
            characterRestrictions: alphaNumericAndHyphen,
            allowedCharacterTypes: alphaNumericAndHyphenAllowedCharacterTypes
        }
    },
    TinyStacks: {
        iacStackName: {
            minimumLength: 2,
            characterLimit: 64,
            characterRestrictions: alphaNumericAndHyphen,
            allowedCharacterTypes: alphaNumericAndHyphenAllowedCharacterTypes
        }
    }
};
exports.resourceNameRules = resourceNameRules;
